File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        int arr[] = new int[301];
        int score[] = new int[301];

        for (int i = 1; i <= n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        //  초기 값
        score[1] = arr[1];
        score[2] = arr[1] + arr[2];
        score[3] = Math.max(arr[2] + arr[3], arr[1] + arr[3]);

        for (int i = 4; i <= n; i++) {
            // 2계단 밟고 본인에게 도착, 한 계단 밟고 본인에게 도착(그 전엔 반드시 2계단이여야함)
            score[i] = Math.max(score[i - 2] + arr[i], score[i - 3] + arr[i - 1] + arr[i]);
        }
        System.out.println(score[n]);
    }
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
import sys

t = int(sys.stdin.readline())

dp = [0] * (t + 1)
arr = []
for i in range(t):
    arr.append(int(sys.stdin.readline()))

if len(arr) < 3:
    # 계단이 3개 이하인경우 합이 최고임
    print(sum(arr))
else:
    dp[0] = arr[0]
    # 안밟는 경우의 수도 존재하지만 최댓값을 만들어야하니까 다음 발판은 무조건 밟아야함
    dp[1] = arr[0] + arr[1]

    for i in range(2, t):
        # 1 2 4 의 경우 와 2 3 의 경우 중 더 큰 수 비교하기
        dp[i] = max(dp[i - 3] + arr[i - 1] + arr[i], dp[i - 2] + arr[i])
    print(dp[t - 1])